apisix/healthcheck_manager.lua

```diff
@@ -0,0 +1,288 @@
+        core.log.warn("version mismatch for resource: ", resource_path,
+                    " current version: ", checker.version, " requested version: ", resource_ver)
+        return nil  -- version not match
+    end
+    return checker
+end
+
+local timer_create_checker_running = false
+function _M.timer_create_checker()
+    if core.table.nkeys(_M.waiting_pool) == 0 then
+        return
+    end
+
+    local waiting_snapshot = tab_clone(_M.waiting_pool)
+    for resource_path, resource_ver in pairs(waiting_snapshot) do
+        do
+            if find_in_working_pool(resource_path, resource_ver) then
+                core.log.info("resource: ", resource_path,
+                             " already in working pool with version: ",
+                               resource_ver)
+                goto continue
+            end
+            local res_conf = fetch_latest_conf(resource_path)
+            if not res_conf then
+                goto continue
+            end
+            local upstream = res_conf.value.upstream or res_conf.value
+            local new_version = res_conf.modifiedIndex .. tostring(upstream._nodes_ver or '')
+            core.log.warn("checking waiting pool for resource: ", resource_path,
+                    " current version: ", new_version, " requested version: ", resource_ver)
+            if resource_ver ~= new_version then
+                goto continue
+            end
+
+            -- if a checker exists then delete it before creating a new one
+            local existing_checker = _M.working_pool[resource_path]
+            if existing_checker then
+                existing_checker.checker:delayed_clear(10)
+                existing_checker.checker:stop()
+                core.log.info("releasing existing checker: ", tostring(existing_checker.checker))
+            end
+            local checker = create_checker(upstream)
+            if not checker then
+                goto continue
+            end
+            core.log.info("create new checker: ", tostring(checker))
+            add_working_pool(resource_path, resource_ver, checker)
+        end
+
+        ::continue::
+        _M.waiting_pool[resource_path] = nil
+    end
+end
+
+local timer_working_pool_check_running = false
+function _M.timer_working_pool_check()
+    if core.table.nkeys(_M.working_pool) == 0 then
+        return
+    end
+
+    local working_snapshot = tab_clone(_M.working_pool)
+    for resource_path, item in pairs(working_snapshot) do
+        --- remove from working pool if resource doesn't exist
+        local res_conf = fetch_latest_conf(resource_path)
+        if not res_conf then
+            item.checker:delayed_clear(10)
+            item.checker:stop()
+            core.log.info("try to release checker: ", tostring(item.checker))
+            _M.working_pool[resource_path] = nil
+            goto continue
+        end
+        local current_ver = res_conf.modifiedIndex ..  tostring(res_conf.value._nodes_ver or '')
```

Comment:
`res_conf.modifiedIndex .. tostring(res_conf.value._nodes_ver or '')` this line of code duplicate in at least three places, can we provide a function to wrap it?

Reply:
done